packages/tabs/src/TabTitle.tsx

```diff
@@ -0,0 +1,154 @@
+import React, { useRef, useEffect, useCallback, SetStateAction } from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { HTMLElementProps } from '@leafygreen-ui/lib';
+
+const listTitle = css`
+  display: inline-block;
+  background-color: ${uiColors.white};
+  color: ${uiColors.gray.base};
+  font-weight: bold;
+  font-size: 16px;
+  line-height: 1;
+  // padding-bottom dervied from border height + 5px
+  padding: 3px 20px 8px;
+  cursor: pointer;
+  transition: 150ms color ease-in-out;
+  border: none;
+  background: none;
+  text-decoration: none;
+
+  &:focus {
+    color: ${uiColors.blue.base};
+    outline: none;
+  }
+
+  &:hover:not(:focus) {
+    color: ${uiColors.gray.dark3};
+  }
+`;
+
+interface SharedTabTitleProps {
+  active: boolean;
+  id?: string;
+  disabled?: boolean;
+  dataTabId: string;
+  ariaControls?: string;
+  setFocusedState: React.Dispatch<SetStateAction<Array<string>>>;
+  className?: string;
+  as?: React.ElementType<any>;
+}
+
+interface LinkTabTitleProps extends HTMLElementProps<'a'>, SharedTabTitleProps {
+  href: string;
+}
+
+interface ButtonTabTitleProps
+  extends HTMLElementProps<'button'>,
+    SharedTabTitleProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedTabTitleProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+export type TabTitleProps =
+  | LinkTabTitleProps
+  | ButtonTabTitleProps
+  | CustomElementButtonProps;
+
+function usesCustomElement(
+  props: TabTitleProps,
+): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function usesLinkElement(
+  props: LinkTabTitleProps | ButtonTabTitleProps,
+): props is LinkTabTitleProps {
+  return props.href != null;
+}
+
+function TabTitle(props: TabTitleProps) {
+  const {
+    active,
+    children,
+    className,
+    disabled,
+    dataTabId,
+    onClick,
+    onKeyDown,
+    ariaControls,
+    setFocusedState,
+    as,
+    ...rest
+  } = props;
+
+  const titleRef = useRef<HTMLElement>(null);
+
+  useEffect(() => {
+    if (titleRef.current && active && !disabled) {
```

Comment:
```suggestion
    if (!disabled && active && titleRef.current) {
```

Micro optimization. The way you have it now we'll have to do a property lookup every iteration. 

Here we're just checking the state of the variable.

This is the smallest of perf optimizations that it's almost silly to worry about but it's easy to implement which is why I'm pointing it out.

Also disabled being false is the most likely case so checking for that first should also help speed things up.

Reply:
👍 Makes sense